File: Sources/structly/03_beginner_recursion/Factorial.swift
//
//  Factorial.swift
//  MyCLI
//
//  Created by new on 3/14/26.
//

func factorial(_ n: Int) -> Int64 {
    func f(n: Int, res: Int64) -> Int64 {
        if n == 0 { return 1 }
        if n == 1 { return res }
        return f(n: n-1, res: res * Int64(n))
    }
    return f(n: n, res: 1)
}

File: Sources/structyC/03_beginner_recursion/factorial.cpp
//
//  factorial.cpp
//  MyCLI
//
//  Created by new on 3/14/26.
//

long fact(int n, long res) {
    if (n == 0) {
        return 1;
    }
    if (n == 1) {
        return res;
    }
    return fact(n-1, res * n);
}

long factorial(int n) {
  return fact(n, 1);
}

